File: leetcode-submissions/62_Unique_Paths.cpp
class Solution {
public:
int totalways(int i,int j,int m,int n){
    //base case
    if(i>=m or j>=n){
        return 0;
    }
    if(i==m-1 and j==n-1){
        return 1;
    }
    return (totalways(i+1,j,m,n)+totalways(i,j+1,m,n));
}
    int uniquePaths(int m, int n) {
      return totalways(0,0,m,n);  
    }
};

File: leetcode-submissions/62_Unique_Paths.ts
function uniquePaths(m: number, n: number): number {
    const dp : Array<number[]> = Array.from({length:m+1},()=>Array.from({length:n+1}, ()=>-1));
    return solve(m , n , 0 , 0, dp);
};

const solve=(m :number , n :number , i :number , j:number ,  dp:Array<number[]>) =>{
    if(i==m-1  && j==n-1){
        return 1;

    }
    
    if(i<0 || j<0 || j>=n || i>=m  ){
        return 0 ;
    }
    if(dp[i][j]!=-1){
        return dp[i][j];
    }
    return dp[i][j]=  solve(m , n , i, j+1, dp)+solve(m , n , i+1, j, dp);
}